Adding upgrade contract tutorial

Buddhika-Pathirathna opened the pull request:
### Related links

Related card - https://github.com/casper-network/docs/issues/179 

### Changes

Added the steps to upgrade with two images.


### Notes
Commented on the link to the previous file in sidebar.config.js file and didn't remove the file yet.

Buddhika-Pathirathna commented on source/docs/casper/dapp-dev-guide/tutorials/upgrade-contract.md:
```diff
@@ -0,0 +1,83 @@
+
+
+a) Import required packages to implement the method
+
```
Is it required to show the import code snippet here?

Buddhika-Pathirathna replied:
Added

ipopescu commented on source/docs/casper/dapp-dev-guide/tutorials/upgrade-contract.md:
```diff
@@ -0,0 +1,83 @@
+
+### Step 2. Implementing the generic method
+
+You need to implement the [`add_contract_version`](https://github.com/casper-network/casper-node/blob/9aa22ac6c998c9d2d5288721974266b0fb44fb36/smart_contracts/contract/src/contract_api/storage.rs#L277-L313) generic method with your contract details. This method is executed automatically when creating a brand new contract and it is considered the very first version of the contract. For future cont ...
```
What happens if the contract was created without this entry point? Is it considered "locked" or "not upgradable"? 
If so, you need to mention that the previously created contract needs to have this entry point in the prerequisites.

Also, please add a code snippet as an example implementation or add a link to an example implementation.

Buddhika-Pathirathna replied:
@darthsiroftardis , need your advice on the first question :)

darthsiroftardis replied:
So this is incorrect:
```
You need to implement the [`add_contract_version`](https://github.com/casper-network/casper-node/blob/9aa22ac6c998c9d2d5288721974266b0fb44fb36/smart_contracts/contract/src/contract_api/storage.rs#L277-L313) generic method with your contract details.
```

You dont need to **implement** this function, but just use as part of the contract api, passing in the **new** entrypoints as part of upgrading the contract. The new contract isn't required to have the entrypoints as the previous contract version. However, that would be a breaking change for any users depending on the upgraded contract, but that is the responsibility of the contract author

Buddhika-Pathirathna replied:
Updated according to Karan's feedback

ipopescu commented on source/docs/casper/dapp-dev-guide/tutorials/upgrade-contract.md:
```diff
@@ -0,0 +1,83 @@
+You need to implement the [`add_contract_version`](https://github.com/casper-network/casper-node/blob/9aa22ac6c998c9d2d5288721974266b0fb44fb36/smart_contracts/contract/src/contract_api/storage.rs#L277-L313) generic method with your contract details. This method is executed automatically when creating a brand new contract and it is considered the very first version of the contract. For future cont ...
+
+
+a) Import required packages to implement the method
```
```suggestion
a) Import required packages to implement the method.
```

ipopescu commented on source/docs/casper/dapp-dev-guide/tutorials/upgrade-contract.md:
```diff
@@ -0,0 +1,83 @@
+
+### Step 4. Deploying the contract 
+
+Use the below command to deploy your contract to the selected network. Make sure you add your local arguments to the command. Refer the [deploying contracts](/docs/dapp-dev-guide/deploying-contracts/) guide to learn more about contract deployment.
```
```suggestion
Use the below command to deploy your contract to the selected network. Make sure you add your specific arguments to the command. Refer to the [deploying contracts](/docs/dapp-dev-guide/deploying-contracts/) guide to learn more about contract deployment.
```

ipopescu commented on source/docs/casper/dapp-dev-guide/tutorials/upgrade-contract.md:
````diff
@@ -0,0 +1,83 @@
+    --session-path [WASM_FILE_PATH]/[File_Name].wasm
+```
+
+- [NETWORK_NAME} = Selected network name
+- [PATH_TO_YOUR_KEY] = Path to your saved keys' location
+- [WASM_FILE_PATH]/[File_Name] - Path to your contract Wasm 
````
Make this list consistent with the list in other parts of the docs (see https://casper.network/docs/workflow/two-party-multi-sig/). Also, there's no need for square brackets. Explain all arguments.

ipopescu reviewed (changes requested):
@Buddhika-Pathirathna @darthsiroftardis please add a few more details in this PR:
1. A brief introduction to upgrades, maybe 3-4 sentences explaining the requirements for a contract to be upgraded (unless you're planning to add this somewhere else). Make sure the prerequisites reflect this. See specific comments in the PR.
2. Usage code snippets or links to an example.
3. Please simplify the images. My suggestions are below and I can help if I get access to the image source.
After reading this tutorial, I am not clear what I need to do exactly to upgrade my counter contract and if I can even do so without having specified the add_contract_version in advance. 

[static/image/addContractVersion.png](https://github.com/casper-network/docs/pull/290/files#diff-a9b851f692cc9ad70d1058aa8f82cb9d481c0d52141c1334200f5fc0f51b48b2)
- I would remove the colors in the boxes, to make it easier to read; keeping it plain is ok. 
- I would have all the steps on one line for easier visual processing
- The arrow from Deploy to .wasm should point in the other direction?

[static/image/contract-representation.png](https://github.com/casper-network/docs/pull/290/files#diff-926b8a7e1af70357029123323bd4bfbbe204167660b5a5a43f887bdad5e7ffa0)
- I suggest to remove italics and colors, keeping the diagrams plain 
- Use “Contract Name” instead of “Contract A” and remove the note at the bottom
- Just say “Contract Package”, instead of “Contract A Package”.

Buddhika-Pathirathna commented on source/docs/casper/dapp-dev-guide/tutorials/upgrade-contract.md:
```diff
@@ -0,0 +1,83 @@
+<img src={useBaseUrl("/image/addContractVersion.png")} alt="add-contract-version" width="600"/>
+
+### Step 1. Creating a new contract file 
+a) Create a new smart contract file that contains the base contract
```
@darthsiroftardis , could you please answer the first question (Does the contract have the same name? I think this is important to mention.) - As per your example, the new contract needs a somewhat very similar name, right?

Buddhika-Pathirathna replied:
Updated according to Karan's feedback

ipopescu commented on config/sidebar.config.js:
```diff
@@ -132,7 +132,8 @@
                     ],
                 },
-                "dapp-dev-guide/tutorials/upgrade-tutorial",
+                // "dapp-dev-guide/tutorials/upgrade-tutorial",
```
@Buddhika-Pathirathna this file has conflicts that need resolution. Please take a look. Also, it's probably best to remove the unused line from here instead of commenting it out.

Buddhika-Pathirathna replied:
Sure, Will work on it

ipopescu commented:
@Buddhika-Pathirathna @darthsiroftardis

I played with this [rudimentary counter contract](https://github.com/casper-ecosystem/counter/pull/17/files#diff-d70ed31cf707bb42d2418f66d3390ef53971f3ae492d33797cbf7e440bde36a7) to understand the versioning requirements. Maybe the code I added there would help with this tutorial.

When talking about the code in this tutorial, you could point to specific lines in the[ counter-define contract](https://github.com/casper-network/casper-node/blob/dev/smart_contracts/contracts/client/counter-define/src/main.rs) under casper-node.

I recommend mentioning the following points -- please check if they're accurate:

- As a prerequisite, the contract writer needs to have a handle on the contract package hash. If there is another way to do this, please explain how. My example code:
`// Create a contract package for this contract and save its hash value.
    let (contract_package_hash, _): (ContractPackageHash, URef) =
    storage::create_contract_package_at_hash();`
- Maybe the name of the contract can change with a different version, but this needs to be clarified. I am still not sure about it.
- When mentioning the add_contract_version implementation, you could point to this code:
`let (stored_contract_hash, contract_version) =
        storage::add_contract_version(contract_package_hash, counter_entry_points, counter_named_keys);`
- On the last step for verifying the new version changes, a test example would be ideal, maybe point [here](https://github.com/casper-network/casper-node/smart_contracts/contracts/test/contract-context/src/main.rs)

I added Adam and Gloreen to have one writer do a pass at the end for flow and comprehension. Thanks!

Buddhika-Pathirathna commented:
> @Buddhika-Pathirathna @darthsiroftardis
> 
> I played with this [rudimentary counter contract](https://github.com/casper-ecosystem/counter/pull/17/files#diff-d70ed31cf707bb42d2418f66d3390ef53971f3ae492d33797cbf7e440bde36a7) to understand the versioning requirements. Maybe the code I added there would help with this tutorial.
> 
> When talking about the code in this tutorial, you could point to specific lines in the[ counter-define contract](https://github.com/casper-network/casper-node/blob/dev/smart_contracts/contracts/client/counter-define/src/main.rs) under casper-node.
> 
> I recommend mentioning the following points -- please check if they're accurate:
> 
> * As a prerequisite, the contract writer needs to have a handle on the contract package hash. If there is another way to do this, please explain how. My example code:
>   `// Create a contract package for this contract and save its hash value. let (contract_package_hash, _): (ContractPackageHash, URef) = storage::create_contract_package_at_hash();`
> * Maybe the name of the contract can change with a different version, but this needs to be clarified. I am still not sure about it.
> * When mentioning the add_contract_version implementation, you could point to this code:
>   `let (stored_contract_hash, contract_version) = storage::add_contract_version(contract_package_hash, counter_entry_points, counter_named_keys);`
> * On the last step for verifying the new version changes, a test example would be ideal, maybe point [here](https://github.com/casper-network/casper-node/smart_contracts/contracts/test/contract-context/src/main.rs)
> * A few more details can be added here: https://casper.network/docs/dapp-dev-guide/upgrading-contracts/. Explain something along these lines: "The contract package hash is like a container, with different versions of the contract, similar to an npm package or a Rust crate. It has several versions of software (v1, v2, v3) with functionality that can differ in certain versions. The contract package is created automatically when you install the contract on the blockchain. The upgrade workflow is essentially adding new versions to the contract package. It is not an in-place replacement and the "upgrade" process is additive. So when we upgrade the contract, we simply add a new version of the contract in the contract package. (This is a key point).The original version of the contract is still there. The developer can enable specific versions of the contract. You can disable a contract version if needed." These are my notes, so they might be useful.
> 
> I added Adam and Gloreen to have one writer do a pass at the end for flow and comprehension. Thanks!

Thanks @ipopescu for taking time for this, I will further go through and try to incorporate the mentioned points.

ipopescu commented:
I feel differently, @darthsiroftardis. The content should be improved and go out asap, and then be further improved with #289. If we place it back in the backlog, it won't be delivered earlier than a month from now, with the current sprint assignments.
If that's not possible, we should remove the current upgrade content completely, because it is incorrect.
It does not need too many updates. If you need my help, let me know, please.

Buddhika-Pathirathna commented on source/docs/casper/dapp-dev-guide/tutorials/upgrade-contract.md:
```diff
@@ -0,0 +1,130 @@
+
+b) Then, add the required modification to the contract. This can be adding a new entry point or modifying the behavior of an existing entry point in the contract.
+
+### Step 2. Add handle to the contract package hash
```
Added this new step to highlight the contract package hash handle

Buddhika-Pathirathna commented on source/docs/casper/dapp-dev-guide/tutorials/upgrade-contract.md:
```diff
@@ -0,0 +1,130 @@
+
+<img src={useBaseUrl("/image/contract-representation.png")} alt="contract-representation" width="500"/>
+
+#### Sample Contract Implementation
```
This part can be removed I guess, sorry couldn't check it again.

ipopescu replied:
Yes, this part needs to be removed, as it overlaps with the previous version. Also, it has an incorrect link to localhost.

Buddhika-Pathirathna replied:
Removed

Buddhika-Pathirathna commented on source/docs/casper/dapp-dev-guide/tutorials/upgrade-contract.md:
```diff
@@ -0,0 +1,130 @@
+### Step 6. Verifying the deploy changes 
+
+#### Testing the upgrade contract 
+You can write unit tests to verify the accuracy of the new contract version. Refer to the code in this [file](https://github.com/casper-network/casper-node/blob/dev/smart_contracts/contracts/test/contract-context/src/main.rs) to get an idea about creating the unit test. This file contains the code for building the test contract with the required constants and entry points. It creates the named-ke ...
```
I am not sure whether we need to illustrate this part further with steps?

ipopescu replied:
I would include a link to the deploy verification in the on-chain-contracts.md file, which you need from the latest version of main.

Buddhika-Pathirathna replied:
Included the link.

ipopescu reviewed (changes requested):
Thanks for the updates, @Buddhika-Pathirathna. There is still some more work to be done here:
- Please update your branch. I still see old files in the branch, when I fetch it, such as deploying-contracts.md. 
- Please resolve open conversations and explain how you addressed them/or not.
- Address Karan’s comment: “You dont need to implement this function, but just use as part of the contract api, passing in the new entrypoints as part of upgrading the contract.” Please update any spots that mention “implement add_contract_version…”, including the [contract-upgrade-flow.png](https://github.com/casper-network/docs/pull/290/files#diff-9af58c6edec88877d5812e02da9ce87a44e4324bdf057264ecfd2b663d3e4beb).
- Fix any broken links/broken build. Make sure the build passes before the next review request.
- For [static/image/contract-upgrade-flow.png](https://github.com/casper-network/docs/pull/290/files#diff-9af58c6edec88877d5812e02da9ce87a44e4324bdf057264ecfd2b663d3e4beb) 
we discussed that the last arrow can be deleted (the arrow from .wasm to “Deploy the contract …” box)
- Also, in the “Contract FIle” box, you have a capital I and F. Please say “Contract file” instead to match the rest of the boxes.
- One of the tech writers could also help review the language and flow.

Buddhika-Pathirathna commented:
> Thanks for the updates, @Buddhika-Pathirathna. There is still some more work to be done here:
> 
> * Please update your branch. I still see old files in the branch, when I fetch it, such as deploying-contracts.md.
> * Please resolve open conversations and explain how you addressed them/or not.
> * Address Karan’s comment: “You dont need to implement this function, but just use as part of the contract api, passing in the new entrypoints as part of upgrading the contract.” Please update any spots that mention “implement add_contract_version…”, including the [contract-upgrade-flow.png](https://github.com/casper-network/docs/pull/290/files#diff-9af58c6edec88877d5812e02da9ce87a44e4324bdf057264ecfd2b663d3e4beb).
> * Fix any broken links/broken build. Make sure the build passes before the next review request.
> * For [static/image/contract-upgrade-flow.png](https://github.com/casper-network/docs/pull/290/files#diff-9af58c6edec88877d5812e02da9ce87a44e4324bdf057264ecfd2b663d3e4beb) 
>   we discussed that the last arrow can be deleted (the arrow from .wasm to “Deploy the contract …” box)
> * Also, in the “Contract FIle” box, you have a capital I and F. Please say “Contract file” instead to match the rest of the boxes.
> * One of the tech writers could also help review the language and flow.
> 
> <img alt="Screen Shot 2022-04-01 at 6 29 14 PM" width="497" src="https://user-images.githubusercontent.com/4185994/161309107-34f1642d-a38b-4531-89e6-04247454a30e.png">

<html xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:dt="uuid:C2F41010-65B3-11d1-A29F-00AA00C14882"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=OneNote.File>
<meta name=Generator content="Microsoft OneNote 15">
</head>

<body lang=en-US style='font-family:Candara;font-size:14.0pt'>
<!--StartFragment-->

<div style='direction:ltr'>



Please         update your branch. I still see old files in the branch, when I fetch         it, such as deploying-contracts.md          - Done     Please         resolve open conversations and explain how you addressed them/or not - Done - Not all explained.     Address         Karan’s comment: “You dont need to implement this function,         but just use as part of the contract api, passing in         the new entrypoints as part of upgrading the contract.”  - Done     Please update any spots that mention “implement         add_contract_version…”, including the contract-upgrade-flow.png. - Done     Fix         any broken links/broken build. Make sure the build passes before the         next review request. - Done     For static/image/contract-upgrade-flow.png                  we discussed that the last arrow can be deleted (the arrow from         .wasm to “Deploy the contract …” box) - Done     Also,         in the “Contract FIle” box, you have a capital I and F. Please say         “Contract file” instead to match the rest of the boxes. - Done     One         of the tech writers could also help review the language and flow. - Done - Have added to the review
--
 



</div>

<!--EndFragment-->
</body>

</html>


• Please update your branch. I still see old files in the branch, when I fetch it, such as deploying-contracts.md  **- Done**
• Please resolve open conversations and explain how you addressed them/or not **- Done** 
• Address Karan’s comment: “You dont need to implement this function, but just use as part of the contract api, passing in the new entrypoints as part of upgrading the contract.”  **- Done**
• Please update any spots that mention “implement add_contract_version…”, including the [contract-upgrade-flow.png](https://github.com/casper-network/docs/pull/290/files#diff-9af58c6edec88877d5812e02da9ce87a44e4324bdf057264ecfd2b663d3e4beb). **- Done**
• Fix any broken links/broken build. Make sure the build passes before the next review request. **- Done**
• For [static/image/contract-upgrade-flow.png](https://github.com/casper-network/docs/pull/290/files#diff-9af58c6edec88877d5812e02da9ce87a44e4324bdf057264ecfd2b663d3e4beb) 
we discussed that the last arrow can be deleted (the arrow from .wasm to “Deploy the contract …” box) **- Done**
• Also, in the “Contract FIle” box, you have a capital I and F. Please say “Contract file” instead to match the rest of the boxes. - **Done**
• One of the tech writers could also help review the language and flow. - **Done - Have added to the review**

ipopescu commented:
@Buddhika-Pathirathna  as you asked me to make any further updates directly, I have made the following changes:

- I updated the introduction and prerequisites to make sure the requirement for having a contract package hash and an unlocked contract is clear
- I removed the contract-upgrade-flow.png. It needed updates and it’s harder to maintain than a short numbered list; plus I don’t have access to the source of the diagram. Updates to note: the use of deploy vs. deploying; build=compile; use add_contract_version instead of implement this function
- I updated the contract versioning flow based on my understanding; I added links to the APIs and updated the explanations => this was a big change
- I moved the general explanation to the landing page for upgrades
- Linked to content already written about deploys, not including it here again
- Storing the version is optional, so I re-wrote and simplified that section 
- **When ready, we need to rebase this branch to dev**

@darthsiroftardis, could you please do another technical review. Thank you, both!

goral09 commented on source/docs/casper/dapp-dev-guide/tutorials/upgrade-contract.md:
```diff
@@ -0,0 +1,114 @@
+  - `entry_points` - Entry points of the contract, which can be modified or newly added
+  - `named_keys` - Named key pairs of the contract
+
+:::note
+
+A few notes to consider:
+
+- The contract file name could differ from the base contract since the contract package hash connects the contract's versions after compiling to Wasm
+- You need to decide how to manage contract versioning with clients using older versions
+
+:::
```
This doesn't render correctly for me:
![image](https://user-images.githubusercontent.com/5120801/162919977-6b151617-ee61-447a-9630-f12352e17a03.png)

ipopescu replied:
If you’re viewing the file in GitHub, it won’t render. I ran the site locally and it was ok. Buddhika or I can double check it on the staging site or locally before the merge.

ipopescu replied:
The note looks ok:

<img width="782" alt="Screen Shot 2022-04-13 at 2 40 19 PM" src="https://user-images.githubusercontent.com/4185994/163181948-cb01011e-77e4-4a4c-a809-b7acdcf8ef28.png">

Buddhika-Pathirathna replied:
The note section renders fine with my localhost.

mpapierski commented on source/docs/casper/dapp-dev-guide/tutorials/upgrade-contract.md:
````diff
@@ -0,0 +1,114 @@
+
+```rust
+    // Create a contract package for this contract and save its hash value
+    let (contract_package_hash, _): (ContractPackageHash, URef) =
````
Discarding access key URef forbids further upgrades

ipopescu replied:
Thanks @mpapierski. I corrected the code and added a new section for locked contract packages with commit 3a0b674bfd46575d964d5ccd4a2c610919cb456b.

ipopescu commented on source/docs/casper/dapp-dev-guide/tutorials/upgrade-contract.md:
````diff
@@ -0,0 +1,114 @@
+
+```rust
+    // Create a contract package for this contract and save its hash value
+    let (contract_package_hash, _): (ContractPackageHash, URef) =
````
I need some help with this explanation. Is it because the package hash is enough?

ipopescu replied:
Addressed with commit 3a0b674bfd46575d964d5ccd4a2c610919cb456b.

mpapierski commented on source/docs/casper/dapp-dev-guide/tutorials/upgrade-contract.md:
````diff
@@ -0,0 +1,114 @@
+
+```rust
+    let (contract_hash, contract_version) = 
+    storage::add_contract_version(contract_package_hash, 
````
You need access key URef (returned from create_contract_package_at_hash) in your context - 
In other words in your named keys. Otherwise you cant add a version for security reasons

ipopescu replied:
Addressed with commit 3a0b674bfd46575d964d5ccd4a2c610919cb456b.

ipopescu commented on source/docs/casper/dapp-dev-guide/tutorials/upgrade-contract.md:
```diff
@@ -0,0 +1,114 @@
+
+Here is an example workflow for creating a versioned contract. Your workflow may differ if you have already created a versioned contract and have a handle on the hash where it is stored in global state.
+
+1. Create a contract that can be versioned
```
Sounds good. I’ll make this update and check for other occurrences as well.

ipopescu replied:
Cleaned up the language with commit 2b76877b7c1ca07f9a09fe230f9d619b6506e491, being careful to say that only the contract package can be versioned (not the contract).

mpapierski commented on source/docs/casper/dapp-dev-guide/tutorials/upgrade-contract.md:
```diff
@@ -0,0 +1,114 @@
+
+  - `contract_package_hash` - This hash directs you to the contract package. See [Hash and Key Explanations](/dapp-dev-guide/understanding-hash-types#hash-and-key-explanations)
+  - `entry_points` - Entry points of the contract, which can be modified or newly added
+  - `named_keys` - Named key pairs of the contract
```
Worth mentioning that we carry on named keys from previous version of the contract. If you specify Some(named_keys) then old named keys will be combined tohether. If new named keys contains keys that also exists in old version then the new values would be present in new version of the contract.

ipopescu replied:
Addressed with commit 16e29b29eed26bc60338d0f09e9763f6510a3e86 in the note section.

mpapierski commented on source/docs/casper/dapp-dev-guide/tutorials/upgrade-contract.md:
```diff
@@ -0,0 +1,114 @@
+A few notes to consider:
+
+- The contract file name could differ from the base contract since the contract package hash connects the contract's versions after compiling to Wasm
+- You need to decide how to manage contract versioning with clients using older versions
```
We can expand on that? For example you can pin your client contract to a contract hash of specific version, or call_versioned_contract with "None" version which will always use newest version of the contract. Or combine call_versioned_contract with a version number to pin to a version.

mpapierski commented on source/docs/casper/dapp-dev-guide/tutorials/upgrade-contract.md:
```diff
@@ -0,0 +1,114 @@
+
+A few notes to consider:
+
+- The contract file name could differ from the base contract since the contract package hash connects the contract's versions after compiling to Wasm
```
I'm not sure I understand - on the network there are no file names after deploying a smart contracts. Is it about Cargo.toml and the name of the Wasm there?

ipopescu replied:
This point refers to the wasm filename. I have updated the explanation with commit 69183474fb6277cac15a779b7986ba1080066b71 to say:
"The Wasm file name of the new contract could differ from the old contract; after sending the deploy to the network, the contract package hash connects the different contract versions".
I hope this is clearer now.

mpapierski reviewed (commented):
Few notes but looks good so far. I think its also worth mentioning (or linking to) 4 modes of calling a contract - contract by named key, contract by hash, versioned contract by package hash, versioned contract by named key (that refers to package hash)

ipopescu commented:
Thanks for your help, @mpapierski, and @goral09. 

I think it's best to work on a separate PR for the details on how to call the upgraded contract as part of card #349. When we revamp the "calling contracts" page, I will link to it as suggested. I have addressed all other feedback.

@darthsiroftardis please give it another look and hopefully we can merge it before Easter ;).

ipopescu commented on docusaurus.config.js:
```diff
@@ -21,7 +21,7 @@ const routePrefix = !docsOnlyMode ? "/docs" : "/";
     ...siteConfig,
     i18n: i18nConfig,
-    baseUrl: "/",
+    baseUrl: '/',
```
Setting this to what we have in the dev branch.

ipopescu commented:
@darthsiroftardis can we merge this [contract-upgrade-tutorial](https://github.com/Buddhika-Pathirathna/docs/tree/contract-upgrade-tutorial)?

ipopescu commented:
As discussed in stand-up, I will merge this PR. Further changes will be added, including examples of calling a specific contract version (see #363 and #349).
